actors/builtin/payment_channel/payment_channel_actor.go

```diff
@@ -0,0 +1,247 @@
+package payment_channel
+
+import (
+	"bytes"
+
+	addr "github.com/filecoin-project/go-address"
+	abi "github.com/filecoin-project/specs-actors/actors/abi"
+	big "github.com/filecoin-project/specs-actors/actors/abi/big"
+	"github.com/filecoin-project/specs-actors/actors/builtin"
+	vmr "github.com/filecoin-project/specs-actors/actors/runtime"
+	indices "github.com/filecoin-project/specs-actors/actors/runtime/indices"
+	"github.com/filecoin-project/specs-actors/actors/serde"
+	"github.com/minio/blake2b-simd"
+)
+
+type PaymentChannelActor struct{}
+
+// type PaymentInfo struct {
+// 	PayChActor     address.Address
+// 	Payer          address.Address
+// 	ChannelMessage *cid.Cid
+
+// 	Vouchers []*types.SignedVoucher
+// }
+
+/////////////////
+// Constructor //
+/////////////////
+
+type PCAConstructorParams struct {
+	To addr.Address
+}
+
+func (pca *PaymentChannelActor) Constructor(rt vmr.Runtime, params *PCAConstructorParams) *vmr.EmptyReturn {
+
+	var st PaymentChannelActorState
+	rt.State().Transaction(&st, func() interface{} {
+		st.From = rt.ImmediateCaller()
+		st.To = params.To
+		st.LaneStates = make(map[int64]*LaneState)
+		return nil
+	})
+	return &vmr.EmptyReturn{}
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Payment Channel state operations
+////////////////////////////////////////////////////////////////////////////////
+
+type PCAUpdateChannelStateParams struct {
+	Sv     SignedVoucher
+	Secret []byte
+	Proof  []byte
+}
+
+func hash(b []byte) []byte {
+	s := blake2b.Sum256(b)
+	return s[:]
+}
+
+type PaymentVerifyParams struct {
+	Extra []byte
+	Proof []byte
+}
+
+func (pca *PaymentChannelActor) UpdateChannelState(rt vmr.Runtime, params *PCAUpdateChannelStateParams) *vmr.EmptyReturn {
+
+	var st PaymentChannelActorState
+	rt.State().Transaction(&st, func() interface{} {
+
+		sv := params.Sv
+
+		vb, nerr := sv.SigningBytes()
+		if nerr != nil {
+			rt.Abort(1, "failed to serialize signedvoucher")
+		}
+
+		if !rt.Syscalls().VerifySignature(*sv.Signature, st.From, vb) {
+			rt.Abort(2, "voucher signature invalid")
+		}
+
+		if rt.CurrEpoch() < sv.TimeLock {
+			rt.Abort(3, "cannot use this voucher yet!")
+		}
+
+		if len(sv.SecretPreimage) > 0 {
+			if !bytes.Equal(hash(params.Secret), sv.SecretPreimage) {
+				rt.Abort(4, "incorrect secret!")
+			}
+		}
+
+		if sv.Extra != nil {
+
+			_, code := rt.Send(
```

Comment:
Illegal to Send() while inside a transaction, the transaction must finish first. Nothing prior to here mutates state, so I suggest using Readonly instead